### PR TITLE
melonDS DS Migration

### PR DESCRIPTION
**Pull request by ghost**
New melonDS RetroArch core: https://github.com/JesseTG/melonds-ds

* Copied old melonDS saves to new melonDSDS
* Copied old melonDS remaps to new melonDSDS
* Added new "Nintendo DS - RetroArch melonDS DS" Steam ROM Manager parser
* Added "(legacy)" to old melonDS Steam ROM Manager parser
* Added melonDSDS function to match new configuration files
* Updated microphone driver to SDL
* No file deletion, legacy melonDS (saves and remaps) still left in tact

### CODE REVIEW

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 1]
```diff
@@ -1541,6 +1543,35 @@ RetroArch_melonDS_setConfig(){
 	RetroArch_setOverride 'nds.cfg' 'melonDS'  'rewind_enable' '"false"'
 }
 
+RetroArch_melonDSDS_setUpCoreOpt(){
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_audio_bitrate' '"Automatic"'
```
Comment (JesseTG) [inline functions/EmuScripts/emuDeckRetroArch.sh:1562]: Equivalent to `melonds_audio_bitdepth = "auto"`, which is the default behavior.

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 2]
```diff
@@ -1541,6 +1543,35 @@ RetroArch_melonDS_setConfig(){
...
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_audio_interpolation' '"None"'
```
Comment (JesseTG) [inline functions/EmuScripts/emuDeckRetroArch.sh:1563]: Equivalent to `melonds_audio_interpolation = "disabled"`, which is the default behavior.

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 3]
```diff
@@ -1541,6 +1543,35 @@ RetroArch_melonDS_setConfig(){
...
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_boot_directly' '"enabled"'
```
Comment (JesseTG) [inline functions/EmuScripts/emuDeckRetroArch.sh:1564]: Equivalent to `melonds_boot_mode = "direct"`, which is the default behavior.

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 4]
```diff
@@ -1541,6 +1543,35 @@ RetroArch_melonDS_setConfig(){
...
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS' 'melonds_console_mode' '"DS"'
```
Comment (JesseTG) [inline functions/EmuScripts/emuDeckRetroArch.sh:1565]: Equivalent to `melonds_console_mode = "ds"`, which is the default.

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 5]
```diff
@@ -1541,6 +1543,35 @@ RetroArch_melonDS_setConfig(){
...
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_dsi_sdcard' '"disabled"'
```
Comment (JesseTG) [inline functions/EmuScripts/emuDeckRetroArch.sh:1566]: The key is unchanged, but the new default value is `enabled`.

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 6]
```diff
@@ -1541,6 +1543,35 @@ RetroArch_melonDS_setConfig(){
...
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_hybrid_ratio' '"2"'
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_hybrid_small_screen' '"Duplicate"'
```
Comment (JesseTG) [inline functions/EmuScripts/emuDeckRetroArch.sh:1568]: Equivalent to `melonds_hybrid_small_screen = "both"`, which is the new default.

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 7]
```diff
@@ -1541,6 +1543,35 @@ RetroArch_melonDS_setConfig(){
...
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_jit_block_size' '"32"'
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_jit_branch_optimisations' '"enabled"'
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_jit_enable' '"enabled"'
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_jit_fast_memory' '"enabled"'
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_jit_literal_optimisations' '"enabled"'
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_opengl_better_polygons' '"enabled"'
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_opengl_filtering' '"nearest"'
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_opengl_renderer' '"enabled"'
```
Comment (JesseTG) [inline functions/EmuScripts/emuDeckRetroArch.sh:1576]: Equivalent to `melonds_render_mode = "opengl"`, but the `software` renderer is actually better in most cases. The software renderer is faster and more accurate, and I recommend that it be used as the default. The OpenGL renderer's chief advantage is its ability to upscale 3D graphics, although it doesn't do so correctly in scenes that use 3D on both screens.

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 8]
```diff
@@ -1541,6 +1543,35 @@ RetroArch_melonDS_setConfig(){
...
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_opengl_resolution' '"5x native (1280x960)"'
```
Comment (JesseTG) [inline functions/EmuScripts/emuDeckRetroArch.sh:1577]: Equivalent to `melonds_opengl_resolution = "5"`

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 9]
```diff
@@ -1541,6 +1543,35 @@ RetroArch_melonDS_setConfig(){
...
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_randomize_mac_address' '"disabled"'
```
Comment (JesseTG) [inline functions/EmuScripts/emuDeckRetroArch.sh:1578]: There is no equivalent for this setting in melonDS DS, the MAC address can't be customized at this time.

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 10]
```diff
@@ -1541,6 +1543,35 @@ RetroArch_melonDS_setConfig(){
...
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_screen_gap' '"0"'
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_screen_layout' '"Hybrid Bottom"'
```
Comment (JesseTG) [inline functions/EmuScripts/emuDeckRetroArch.sh:1580]: There is no single equivalent to this setting, as up to 8 screen layouts can be used in sequence.

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 11]
```diff
@@ -1541,6 +1543,35 @@ RetroArch_melonDS_setConfig(){
...
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_swapscreen_mode' '"Toggle"'
```
Comment (JesseTG) [inline functions/EmuScripts/emuDeckRetroArch.sh:1581]: There is currently no equivalent to this setting in melonDS DS, although I may implement one in the future.

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 12]
```diff
@@ -1541,6 +1543,35 @@ RetroArch_melonDS_setConfig(){
...
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_threaded_renderer' '"disabled"'
```
Comment (JesseTG) [inline functions/EmuScripts/emuDeckRetroArch.sh:1582]: I would default this setting to `enabled`, as the threaded renderer is now stable enough to do so.

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 13]
```diff
@@ -1541,6 +1543,35 @@ RetroArch_melonDS_setConfig(){
...
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_touch_mode' '"Touch"'
```
Comment (JesseTG) [inline functions/EmuScripts/emuDeckRetroArch.sh:1583]: Equivalent to `melonds_touch_mode = "pointer"`, though I recommend a default value of `auto`. `auto` allows the touch screen to be used with the pointer, touch screen, or joystick (depending on which was last used).

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 14]
```diff
@@ -1541,6 +1543,35 @@ RetroArch_melonDS_setConfig(){
 	RetroArch_setOverride 'nds.cfg' 'melonDS'  'rewind_enable' '"false"'
 }
 
+RetroArch_melonDSDS_setUpCoreOpt(){
```
[the remark this answers is not in the document]
Comment (JesseTG) [reply functions/EmuScripts/emuDeckRetroArch.sh:1561]: Any setting that I _didn't_ annotate will be accepted unchanged, although they represent defaults that are already in place.

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 15]
```diff
@@ -2343,6 +2374,38 @@ RetroArch_autoSave(){
...
+if [ ! -d "$melondsds_config" ]; then
+  mkdir -p "$melondsds_config"
+fi
+
+if [ ! -d "$melondsds_remaps" ]; then
+  mkdir -p "$melondsds_remaps"
+fi
+
+cp -r "$melonds_config/." "$melondsds_config"
+mv "$melondsds_config/melonDS.opt" "$melondsds_config/melonDS DS.opt"
```
[the remark this answers is not in the document]
Comment (JesseTG) [reply functions/EmuScripts/emuDeckRetroArch.sh:2419]: Note that if RetroArch parses a known option key with an unknown option value, the parsed value will be discarded in favor of the core's default.

**Review by JesseTG — CHANGES_REQUESTED** [review]
I appreciate your hard work on getting this into EmuDeck. With the sensible defaults that I already provide, I don't think you need to do much (if any) core option overrides.

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 16]
```diff
@@ -1556,6 +1558,35 @@ RetroArch_melonDS_setConfig(){
...
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_randomize_mac_address' '"disabled"'
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_screen_gap' '"0"'
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_screen_layout' '"Hybrid Bottom"'
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_swapscreen_mode' '"Toggle"'
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_threaded_renderer' '"disabled"'
+	RetroArch_setOverride 'melonDS DS.opt' 'melonDS DS'  'melonds_touch_mode' '"Touch"'
+}
+
+RetroArch_melonDSDS_setConfig(){
+	RetroArch_setOverride 'nds.cfg' 'melonDS DS'  'rewind_enable' '"false"'
```
Comment (JesseTG) [inline functions/EmuScripts/emuDeckRetroArch.sh:1587]: I would override the default microphone driver to `alsa` or `sdl2`, as there is [a known bug](https://github.com/JesseTG/melonds-ds/issues/134) that I strongly suspect is caused by the default `alsathread` microphone driver. (The audio driver is unaffected.)

File: functions/EmuScripts/emuDeckRetroArch.sh [hunk 17]
```diff
@@ -2359,6 +2399,31 @@ RetroArch_autoSave(){
...
 }
 
+RetroArch_melonDSDSMigration(){
+
+local RetroArch_saves="$RetroArch_path/saves"
+local melonDS_remaps="$RetroArch_path/config/remaps/melonDS"
+local melonDSDS_remaps="$RetroArch_path/config/remaps/melonDS DS"
+
+# Copying melonDS saves to melonDSDS
+for file in $RetroArch_saves/*.sav; do
```
Comment (Godsbane) [inline functions/EmuScripts/emuDeckRetroArch.sh:2409]: wouldn't this rename all .sav files? Are they exclusively melonds core saves?

**Review by Godsbane — APPROVED** [review]
Looks good to me.

If @JesseTG is good with it, then I'd say merge away.

**Comment by JesseTG** [comment]
> Looks good to me.
> 
> If @JesseTG is good with it, then I'd say merge away.

I am. Ship it. :+1:

**Review by JesseTG — APPROVED** [review]

**Pull request merged**